sqlglot/optimizer/eliminate_join_marks.py

```diff
@@ -0,0 +1,184 @@
+import typing as t
+import copy
+
+from sqlglot import expressions as exp
+from sqlglot.optimizer.scope import build_scope, find_all_in_scope
+
+
+def _update_from(
+    select: exp.Select,
+    new_join_dict: t.Dict[str, exp.Join],
+    old_join_dict: t.Dict[str, exp.Join],
+):
+    """If the from clause needs to become a new join, find an appropriate table to use as the new from.
+    updates select in place
+
+    Args:
+        select (exp.Select): The select statement to update
+        new_join_dict (Dict[str, exp.Join]): The dictionary of new joins
+        old_join_dict (Dict[str, exp.Join]): The dictionary of old joins
+    """
+    old_from = select.args["from"]
+    if old_from.alias_or_name not in new_join_dict.keys():
+        return
+    in_old_not_new = old_join_dict.keys() - new_join_dict.keys()
+    if len(in_old_not_new) >= 1:
+        new_from_name = list(old_join_dict.keys() - new_join_dict.keys())[0]
+        new_from_this = old_join_dict[new_from_name].this
+        new_from = exp.From(this=new_from_this)
+        del old_join_dict[new_from_name]
+        select.set("from", new_from)
+    else:
+        raise ValueError("Cannot determine which table to use as the new from")
+
+
+def _update_join_dict(join: exp.Join, join_dict: t.Dict[str, exp.Join]) -> t.Dict[str, exp.Join]:
+    """Update the join dictionary with the new join.
+    If the join already exists, update the on clause.
+
+    Args:
+        join (exp.Join): The join to add to the dictionary
+        join_dict (Dict[str, exp.Join]): dictionary of joins where str is join.alias_or_name
+
+    Returns:
+        Dict[str, exp.Join]: The updated dictionary of joins
+    """
+    if join.alias_or_name in join_dict.keys():
+        join_dict[join.alias_or_name].set(
+            "on",
+            exp.And(
+                this=join_dict[join.alias_or_name].args["on"],
+                expression=join.args["on"],
+            ),
+        )
+    else:
+        join_dict[join.alias_or_name] = join
+    return join_dict
+
+
+def _clean_binary_node(node: exp.Expression):
+    """if the node is left with only one child, promote the child to the parent node.
+    transformation is done in place.
+
+    Args:
+        node (exp.Expression): The node to clean"""
```

Comment:
delete all of these doc string type annotations, the method is already annotated

Reply:
removed